modules/stochastic_tools/src/reporters/AdaptiveMonteCarloDecision.C

```diff
@@ -90,25 +103,45 @@ AdaptiveMonteCarloDecision::execute()
   if (_ais)
   {
     const Real tmp = _ais->getUseAbsoluteValue() ? std::abs(_output_value[0]) : _output_value[0];
-    const bool output_limit_reached = tmp >= _output_limit;
+    bool output_limit_reached;
+
+    /* Checking whether a GP surrogate is used. If it is used, importance sampling is not performed
+    during the training phase of the GP and all proposed samples are accepted until the training
+    phase is completed. Once the training is completed, the importance sampling starts.
+
+    If a GP surrogate is not used, the standard proposal and acceptance/rejection is performed as
+    part of the importance sampling. */
+    bool restart_gp = 0;
+    output_limit_reached = tmp >= _output_limit;
+    if (isParamValid("gp_decision"))
+    {
+      const int training_samples =
+          getUserObject<ActiveLearningGPDecision>("gp_decision").getTrainingSamples();
+      restart_gp = _step == training_samples;
+      output_limit_reached = 1.0;
+      if (restart_gp)
+        reinitChain();
+    }
+
     _output_required[0] = output_limit_reached ? 1.0 : 0.0;
-    if (_step <= _ais->getNumSamplesTrain())
+
+    if (_step <= _ais->getNumSamplesTrain() && !restart_gp)
     {
       /* This is the training phase of the Adaptive Importance Sampling algorithm.
-         Here, it is decided whether or not to accept a proposed sample by the
-         AdaptiveImportanceSampler.C sampler depending upon the model output_value. */
+        Here, it is decided whether or not to accept a proposed sample by the
+        AdaptiveImportanceSampler.C sampler depending upon the model output_value. */
       _inputs = output_limit_reached
                     ? StochasticTools::reshapeVector(_sampler.getNextLocalRow(), 1, true)
                     : _prev_val;
       if (output_limit_reached)
         _prev_val = _inputs;
       _prev_val_out = _output_required;
     }
-    else
+    else if (_step > _ais->getNumSamplesTrain() && !restart_gp)
     {
       /* This is the sampling phase of the Adaptive Importance Sampling algorithm.
-         Here, all proposed samples by the AdaptiveImportanceSampler.C sampler are accepted since
-         the importance distribution traning phase is finished. */
+        Here, all proposed samples by the AdaptiveImportanceSampler.C sampler are accepted since
+        the importance distribution traning phase is finished. */
       _inputs = StochasticTools::reshapeVector(_sampler.getNextLocalRow(), 1, true);
       _prev_val_out[0] = tmp;
```

Comment:
I see a lot of `[0]`, do this and `output_required` need to be vectors?

Reply:
Yes, this class generalizes across all MCMC-based forward UQ methods. AIS has a single Markov chain hence we use [0] but PSS has multiple Markov chains and we need a vector.

Reply:
Nice, thanks for the explanation!